File: php/algo/qsort.php
<?php

require __DIR__.'/vendor/autoload.php';

use function Prelude\{gt, lte, head, tail, filter};

function qsort(array $xss)
{
    if ([] === $xss) {
        return [];
    }

    $pivot = head($xss);
    $xs = tail($xss);

    $greaterThanPivot = filter(gt($pivot));
    $lessThanPivot = filter(lte($pivot));

    return array_merge(
        qsort($greaterThanPivot($xs)),
        [$pivot],
        qsort($lessThanPivot($xs))
    );
}

assert(qsort([2, 1, 4, 100, 77, 3, 99, 5]) === [1, 2, 3, 4, 5, 77, 99, 100]);

File: python/algo/qsort.py
#!/usr/bin/python

def qsort(xss: list) -> list:
    if ([] == xss):
        return []

    pivot = xss[0]
    xs = xss[1:] 

    gt_pivot = lambda xs: filter(lambda y: pivot > y, xs)
    lte_pivot = lambda xs: filter(lambda y: pivot <= y, xs)

    return qsort(list(gt_pivot(xs))) + [pivot] + qsort(list(lte_pivot(xs)))

print(qsort([]))
print(qsort([2, 1, 4, 100, 77, 3, 99, 5]))
